metadata-service/health-servlet/src/main/java/com/datahub/health/controller/HealthCheckController.java

```diff
@@ -26,11 +32,46 @@ public class HealthCheckController {
   @Autowired
   @Qualifier("elasticSearchRestHighLevelClient")
   private RestHighLevelClient elasticClient;
-  private Supplier<ResponseEntity<String>> memoizedSupplier;
+  private final Supplier<ResponseEntity<String>> memoizedSupplier;
+
+  public HealthCheckController(ConfigurationProvider config) {
 
-  public HealthCheckController() {
     this.memoizedSupplier = Suppliers.memoizeWithExpiration(
-        this::getElasticHealth, 5, TimeUnit.SECONDS);
+        this::getElasticHealth, config.getHealthCheck().getCacheDurationSeconds(), TimeUnit.SECONDS);
+  }
+
+  /**
+   * Combined health check endpoint for checking GMS clients.
+   * For now, just checks the health of the ElasticSearch client
+   * @return A ResponseEntity with a Map of String (component name) to ResponseEntity (the health check status of
+   * that component). The status code will be 200 if all components are okay, and 500 if one or more components are not
+   * healthy.
+   */
+  @GetMapping(path = "/ready", produces = MediaType.APPLICATION_JSON_VALUE)
+  public ResponseEntity<Map<String, ResponseEntity<String>>> getCombinedHealthCheck(String... checks) {
+
+    Map<String, Supplier<ResponseEntity<String>>> healthChecks = new HashMap<>();
+    healthChecks.put("elasticsearch", this::getElasticHealthWithCache);
+    // Add new components here
+
+    List<String> componentsToCheck = checks != null && checks.length > 0
+        ? Arrays.asList(checks)
+        : new ArrayList<>(healthChecks.keySet());
+
+    Map<String, ResponseEntity<String>> componentHealth = new HashMap<>();
+    for (String check : componentsToCheck) {
+      componentHealth.put(check,
+          healthChecks.getOrDefault(check,
+              () -> ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body("Unrecognized component " + check))
+              .get());
+    }
+
+
+    boolean isHealthy = componentHealth.values().stream().allMatch(resp -> resp.getStatusCode() == HttpStatus.OK);
+    if (isHealthy) {
+      return ResponseEntity.ok(componentHealth);
+    }
+    return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(componentHealth);
```

Comment:
I like 503 SERVICE_UNAVAILABLE better for these I think, but 500 is okay too :) Just less specific

Reply:
good idea, done